Docker image added

Antoinebr opened the pull request:
### What type of change does this PR introduce?
- [ ] Bugfix
- [X] Feature
- [ ] Refactor
- [X] Documentation
- [ ] Not Sure?

### Does this PR introduce breaking changes?
- [ ] Yes
- [X] No

### List any relevant issue numbers:

https://github.com/leon-ai/leon/issues/1
https://github.com/leon-ai/leon/issues/4

### Description:

I added a DockerImage to run Leon, I also detailed the steps in the ```readme.md```

Unfortunately I saw the leon-ci-images afterward so my image is a bit different (so could be refactored)

gabrielcalderon commented:
Shouldn't this be using the official Node Docker Image instead of a third-party?

https://hub.docker.com/_/node/

Consider using "--no-cache" when adding Alpine packages to reduce final image size.

Consider running "apk upgrade" as the first command to keep the base alpine up-to-date.

louistiti commented:
Hello @Antoinebr,

Thanks for your PR. However, it would be better to:

1. Use the same prerequisites installations as the leon-ci-image [Dockerfile](https://github.com/leon-ai/leon-ci-image/blob/master/Dockerfile). It is minimalist enough.
2. Use an official Node.js image (the same as the leon-ci-image actually).
3. You do not need to install TinyDB as it is already installed during the npm `postinstall` script. Not sure why you did not have TinyDB installed, but you still can run `npm run postinstall` to do so.

@gabrielcalderon feel free to edit his Dockerfile based on the leon-ci-image Dockerfile :smiley:

Antoinebr commented:
@gabrielcalderon @louistiti 

Thank you for your great feedback

I cleaned the Dockerfile to match https://github.com/leon-ai/leon-ci-image/blob/master/Dockerfile 

I also fixed a typo on ```readme.md```  

Let me know your thoughts :)

gabrielcalderon commented:
Looks good to me!

OvermindDL1 commented:
Oh hey, that was fast.  ^.^

For note, if you make a multi stage Dockerfile then you can build everything and prepare in one image, then copy all the needed bits to run it to a clean reduced new image and wipe the old image.  It makes for much smaller distributions and storage requirements.

Antoinebr commented:
> Oh hey, that was fast. ^.^
> 
> For note, if you make a multi stage Dockerfile then you can build everything and prepare in one image, then copy all the needed bits to run it to a clean reduced new image and wipe the old image. It makes for much smaller distributions and storage requirements. It's the Docker builder pattern.

I didn't know that, do you have any documents or articles to help me to understand the concept ? :p

OvermindDL1 commented:
It's quite simple actually, it's like a normal Dockerfile except pretend you are putting multiple Dockerfile's together, so you will have multiple FROM's and all, you can then use the COPY command to copy files from the prior images to the current based on their index definition in the file, and the last specified FROM image is the one that finally gets output.  :-)

An example:
```Dockerfile
FROM my_image:1.2.3

WORKDIR /my/build/dir

RUN compile thing

FROM alpine:latest
RUN apk --no-cache add ca-certificates

WORKDIR /root/

COPY --from=0 /my/build/dir/app.binary    .

CMD ["./app.binary"]
```
Just copy over everything you need to actually run it and nothing more.  An alpine image is often the output image I use, I just install specifically what I need, like the python runtime or node runtime or so, no build systems, etc...  With decent programming languages that can actually output native binaries then generally all you have to copy is the output binary and its data files, no installation of anything is needed or anything of the sort at that point, super-slim.  :-)

The official docs:
https://docs.docker.com/develop/develop-images/multistage-build/

gabrielcalderon commented:
I manage to create a multistage build that reduces the final size to ~800MB from 1.3GB, but it currently fails to start finding modules.

* Running "npm install --only=production", fails. Which makes the node_modules folder massive.
* I tried creating an actual binary using @zeit/pkg and failed. The tool requires an entry point.

This is my dependencies and build stage. 

```
# --- Step 1: Dependencies ---
FROM node:10-alpine AS dependencies

RUN apk update && \
        apk upgrade --no-cache && \
        apk add --no-cache  \
        ca-certificates \
        build-base \
        python3 \
        git && \
        pip3 install --upgrade --no-cache-dir pip && \
        pip3 install --no-cache-dir pipenv

COPY package*.json ./
RUN npm install

# --- Step 2: Build Stage ---
FROM dependencies AS build
COPY . .
RUN npm run build
```

You can then reference the previous stages using

```
COPY --from=dependencies ./node_modules ./nodule_modules/
```

Tchoupinax commented:
Hello,

@Antoinebr I don't know how to commit code on your PR so I will describe my suggestions :

* Python modules have to be installed, `npm run check` verification failed when I tried your Dockerfile. Add **`RUN cd ./bridges/python/ && pipenv install && cd /app`** fix that.
* As far I know, post install script is automatically triggered when installation is finished, why did you add it as a Dockerfile instruction ?

:)

louistiti commented:
![](https://github.trello.services/images/mini-trello-icon.png) [✨ [1.0.0-beta.1] Docker image](https://trello.com/c/VOhSdIC4/215-%E2%9C%A8-100-beta1-docker-image)

DblK commented:
You can also use https://www.npmjs.com/package/modclean to reduce the size of the output docker image by remove unnecessary files in node_modules.

DblK commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+# Install Leon
+RUN npm install
+RUN npm run build
+RUN npm run postinstall
```
Remove this line

Antoinebr replied:
done

Tchoupinax commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+RUN npm install
+RUN npm run build
+RUN npm run postinstall
+
```
add **RUN cd ./bridges/python/ && pipenv install && cd /app** to install Python dependencies

Antoinebr replied:
done

louistiti commented:
Hello all,

Thanks a lot for helping with that :+1: 
I edited few things.

In this Docker environment I noticed that npm is not triggering the preinstall and postinstall scripts because npm tries to downgrade its privileges. So these scripts were never executed, which means Leon was not correctly installed.

Cheers :beers:
